### PR TITLE
VSCE: Implement function to validate current branch

### PR DESCRIPTION
**Pull request by abeatrix**
Close https://github.com/sourcegraph/sourcegraph/issues/34591

# Summary

This PR implements a function to check if the current branch exists on the sourcegraph instance

If the user is currently on a branch that does not exist remotely, the links-commands will open the repo using the "HEAD" (aka main) branch unless a default branch has been set

## Expected Behavior

For local files, users can choose between open / copy the main branch and the current branch.

- If the current branch does not exist remotely, it will fall back to the main branch.
- If the default branch is set, it will replace the current branch option when clicked.

Screen Shot 2022-05-17 at 9 26 58 PM

 

For remote files, only current branch is available:
Screen Shot 2022-05-17 at 9 26 44 PM

## TODO
- [x] Loom video: https://www.loom.com/share/b5e326959696455986cf68088a53b28a
- [x] Test with different remote URL 

## Test plan

Manually tested. See [loom video](https://www.loom.com/share/b5e326959696455986cf68088a53b28a)

## App preview:

- [Web](https://sg-web-bee-vsce-links-commands.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-ksdpikqylk.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/vscode/package.json [hunk 1]
```diff
@@ -47,11 +47,31 @@
...
       },
+      {
+        "command": "sourcegraph.openMain",
+        "category": "Sourcegraph",
+        "title": "Go to main branch"
+      },
+      {
+        "command": "sourcegraph.openCurrent",
+        "category": "Sourcegraph",
+        "title": "Go to current Branch"
```
Comment (tjkandala) [inline client/vscode/package.json:58]:
```suggestion
        "title": "Go to current branch"
```

File: client/vscode/package.json [hunk 2]
```diff
@@ -179,12 +199,14 @@
...
-          "command": "sourcegraph.openInBrowser",
-          "group": "sourcegraph"
+          "submenu": "sourcegraph.openInBranch",
+          "group": "sourcegraph",
+          "label": "sourcegraph"
         },
         {
-          "command": "sourcegraph.copyFileLink",
-          "group": "sourcegraph"
+          "submenu": "sourcegraph.CopyBranchLink",
```
Comment (tjkandala) [inline client/vscode/package.json:207]:
```suggestion
          "submenu": "sourcegraph.copyBranchLink",
```

File: client/vscode/package.json [hunk 3]
```diff
@@ -210,7 +254,17 @@
...
       ]
-    }
+    },
+    "submenus": [
+      {
+        "id": "sourcegraph.openInBranch",
+        "label": "Open File in Sourcegraph Web"
+      },
+      {
+        "id": "sourcegraph.CopyBranchLink",
```
Comment (tjkandala) [inline client/vscode/package.json:264]:
```suggestion
        "id": "sourcegraph.copyBranchLink",
```

File: client/vscode/src/link-commands/git-helpers.ts [hunk 4]
```diff
@@ -41,17 +44,15 @@ export async function repoInfo(filePath: string): Promise<RepositoryInfo | undef
...
         let { branch, remoteName } = await gitRemoteNameAndBranch(repoRoot, gitHelpers, log)
-
-        branch = getDefaultBranch() || branch
-
         const remoteURL = await gitRemoteUrlWithReplacements(repoRoot, remoteName, gitHelpers, log)
-
+        // check if current branch exists on sourcegraph. If not, open in HEAD instead
+        const validatedBranch = await checkBranch(remoteURL, branch)
+        // Use default branch if set. If not, use validated branch instead
+        branch = getDefaultBranch() || validatedBranch
```
Comment (tjkandala) [inline client/vscode/src/link-commands/git-helpers.ts:55]:
```suggestion
        branch = getDefaultBranch() || await checkBranch(remoteURL, branch)
```

We can avoid doing the hard work of validating the branch when there's a default branch through short-circuit evaluation

File: client/vscode/src/link-commands/git-helpers.ts [hunk 5]
```diff
@@ -187,18 +188,21 @@ export function getSourcegraphFileUrl(
...
+        file: encodeURIComponent(fileRelative),
+        editor: encodeURIComponent('VSCode'),
+        version: encodeURIComponent(version),
+        start_row: encodeURIComponent(String(editor.selection.start.line)),
+        start_col: encodeURIComponent(String(editor.selection.start.character)),
+        end_row: encodeURIComponent(String(editor.selection.end.line)),
+        end_col: encodeURIComponent(String(editor.selection.end.character)),
+    }
+    const uri = new URL('/-/editor', SourcegraphUrl)
+    const parametersString = new URLSearchParams({ ...parameters }).toString()
```
Comment (tjkandala) [inline client/vscode/src/link-commands/git-helpers.ts:203]: Nice refactoring!

**Review by tjkandala — APPROVED** [review]

**Comment by abeatrix** [comment]
> I don't think this UX quite makes sense.
> 
> > If the current branch does not exist remotely, it will fall back to the main branch.
> 
> But here we can't guarantee that this code _actually_ exists in main branch right? This is likely not what you wanted. Instead would it make sense after clicking the button, not copy the link to the clipboard and throw an alert saying that the branch isn't published?
> 
> Also, what happens if you use the command palette?

@ryankscott Ohhhh I think I understand your concerns now, thanks for clarifying!

For me when I open the file link in Web, I was hoping it would lead me directly to the file that has the latest code even if my code doesn't exists remotely yet. I think this is what Kelli meant when she said in [slack](https://sourcegraph.slack.com/archives/C03CSAER9LK/p1650987918954939?thread_ts=1650987717.329339&cid=C03CSAER9LK) that:

> 99% of the time I’m doing this, I’m not on main, and I’d rather share the link on main... But almost just as often, I haven’t already published my branch, so I just get an ugly 404 page and end up having to go modify the URL manually

But I see how taking them directly to `main` would not make sense and a warning would be nice.  To address this, do you think we should still open/copy the link to the `main` branch even if the current branch does not exist remotely, and display a warning to let them know the current branch has not been pushed yet?

File: client/vscode/src/link-commands/git-helpers.ts (hunk 5 above)
Comment (abeatrix) [reply client/vscode/src/link-commands/git-helpers.ts:203]: Thank you <3

**Comment by ryankscott** [comment]
> But I see how taking them directly to main would not make sense and a warning would be nice. To address this, do you think we should still open/copy the link to the main branch even if the current branch does not exist remotely, and display a warning to let them know the current branch has not been pushed yet?

I don't think we should open/copy the link but definitely display an alert (we already have one). 
image


It's deterministic that we know that the link will never work right? If that's the case we know we're making it appear that Sourcegraph is broken. The risk that I see is that people don't see / ignore the alert and think the link copying is broken. However, I think this approach is preferable (catching it early, displaying a relevant warning) rather than pushing this to the Sourcegraph web app.

**Comment by abeatrix** [comment]
> > But I see how taking them directly to main would not make sense and a warning would be nice. To address this, do you think we should still open/copy the link to the main branch even if the current branch does not exist remotely, and display a warning to let them know the current branch has not been pushed yet?
> 
> I don't think we should open/copy the link but definitely display an alert (we already have one). image
> 
> It's deterministic that we know that the link will never work right? If that's the case we know we're making it appear that Sourcegraph is broken. The risk that I see is that people don't see / ignore the alert and think the link copying is broken. However, I think this approach is preferable (catching it early, displaying a relevant warning) rather than pushing this to the Sourcegraph web app.

Yes, that makes sense! I have updated the code so that:
- Copy / Open `Main Branch`
  - if `default branch` has been set, use the `default branch`. 
  - If no `default branch`, use `main branch`
- Copy / Open `Current Branch`
  - if `default branch` has been set, use the `default branch`
  - If no `default branch`, use `current branch`
  - If the `current branch` does not exist remotely, display an error and no link will be opened / copied

The command palette will use the `Copy / Open Current Branch` logic as we currently do, but instead of taking them to a 404 page, it will now display an error message telling them the branch does not exist.

Screenshot of the error message when the current branch does not exists and no default branch is set:
Screen Shot 2022-05-19 at 2 26 09 PM


@ryankscott please let me know if this is what you have in mind, thanks!

**Comment by ryankscott** [comment]
>     Copy / Open Main Branch
>         if default branch has been set, use the default branch.
>         If no default branch, use main branch
>     Copy / Open Current Branch
>         if default branch has been set, use the default branch
>         If no default branch, use current branch
>         If the current branch does not exist remotely, display an error and no link will be opened / copied

Cool, do you think we need both options 🤔 ? I'm trying to think of a situation where you are working on a branch, but you want to share it on main (and that code exists there)? Perhaps you've checked out a branch and not made changes?

If there's a legitimate use case for sharing main let's do it. Otherwise, I think we could collapse this to one option that does what your second option proposed.

Depending on the above then the alert should likely give you slightly different feedback. 

"Current branch does not exist on Sourcegraph. Please publish your branch and try again"
or
"Current branch does not exist on Sourcegraph. Please either publish your branch or, copy a link from your main branch."

Thoughts?

**Comment by abeatrix** [comment]
@ryankscott What if we keep just one option, where we will show a popup if we can't find the current branch remotely, and ask if they want to continue using the main branch?

For example:
image

**Comment by ryankscott** [comment]
Yep I think that works, nice one! 
Perhaps we can get @jjinnii to take a quick look and give it a stamp, now she's back.

**Comment by abeatrix** [comment]
Yes you got it! Will update the copy in a bit, thanks for taking a look Sara 😀

**Pull request merged**